Improve Vala support

al1-ce opened the pull request:
When compiling multiple files valac fails to find symbols (i.e SomeClass from a.vala using in b.vala), changed build to process all source files at once

waruqi commented on xmake/rules/vala/xmake.lua:
```diff
@@ -58,67 +58,82 @@ rule("vala.build")
+
+        for _, sourcefile_c in ipairs(sourcefiles_c) do
+            batchcmds:show_progress(opt.progress, "${color.build.object}compiling.c %s", sourcefile_c)
+            batchcmds:compile(sourcefile_c, target:objectfile(sourcefile_c))
```
Using before_buildcmd_files will not compile them in parallel.

al1-ce replied:
The problem is that you can't compile vala code separately

Imagine simple example:
```vala
// main.vala
int main() {
    myfunc(); // valac will throw error on this line
}
```

```vala
// b.vala
void myfunc() { stdout.print("impossible"); }
```

Vala compiler has no way of knowing if there's a `myfunc` defined anywhere if compiling files separately, that's why I'm making this PR

waruqi replied:
But here you are also compiling them one by one instead of passing a list of source files to vala all at once, so how does vala know t their symbols?

Can you provide a full test project as attachment? I will look at it.

al1-ce replied:
I was and am passing all vala files to compiler by adding them to argv

But, I refactored it and vala still has to go through entire project, but generated C code should be incrementally build now (and now Vala can properly directly import C symbols)

For project look at `tests/projects/vala/includec/src/main.vala` from my latest commit

waruqi commented on xmake/rules/vala/xmake.lua:
```diff
@@ -58,67 +58,82 @@ rule("vala.build")
+        end
+        for _, objectfile in ipairs(target:objectfiles()) do
+            batchcmds:set_depmtime(os.mtime(objectfile))
+            batchcmds:set_depcache(target:dependfile(objectfile))
```
This will break incremental compilation.

al1-ce replied:
Is there a way to set batchcmd depend from hash instead of modify time if to move compilation of **only** outputed C code into before_buildcmd_file?

al1-ce replied:
Doesn't matter anymore, see latest commit

waruqi commented on xmake/rules/vala/xmake.lua:
```diff
@@ -58,31 +56,40 @@ rule("vala.build")
         end
     end)
-    before_buildcmd_file(function (target, batchcmds, sourcefile_vala, opt)
+    before_buildcmd_files(function (target, batchcmds, sourcefiles_vala, opt)
```
please rename `sourcefiles_vala` to `sourcebatch`

waruqi commented on xmake/rules/vala/xmake.lua:
```diff
@@ -58,31 +56,40 @@ rule("vala.build")
+        -- otherwise valac would fail when compiling multiple files
+        for _, sourcefile_vala in ipairs(sourcefiles_vala.sourcefiles) do
+            -- if it's only a vala file
+            if string.match(sourcefile_vala, ".vala$") ~= nil then
```
use path.extension

waruqi commented on xmake/rules/vala/xmake.lua:
```diff
@@ -103,22 +110,39 @@ rule("vala.build")
+
+    on_buildcmd_file(function (target, batchcmds, sourcefile_vala, opt)
+        -- Again, only vala files need special treatment
+        if string.match(sourcefile_vala, ".vala$") ~= nil then
```
and here

waruqi commented on xmake/rules/vala/xmake.lua:
```diff
@@ -103,22 +110,39 @@ rule("vala.build")
+            batchcmds:set_depcache(target:dependfile(objectfile))
+
+            batchcmds:show_progress(opt.progress, "${color.build.object}compiling.c %s", sourcefile_c)
+            batchcmds:compile(sourcefile_c, objectfile, { configs = { force = { cflags = "-w" } } })
```
why add `-w`? we can use `set_warnings` in project xmake.lua

al1-ce replied:
Because there's expected warnings from files that are generated from Vala and it's [recommended to disable them by Vala devs](http://docs.vala.dev/tutorials/programming-language/main/07-00-tools/07-01-valac.html):
```
The recommended approach is to just disable those warnings by passing options to the C compiler:
valac -X -w hello.vala
```
And this option only going to touch valac generated C files

waruqi commented:
Did you test other examples.

e.g. https://github.com/xmake-io/xmake/tree/dev/tests/projects/vala/staticlib

it does not work.

and others. https://github.com/xmake-io/xmake/tree/dev/tests/projects/vala

al1-ce commented:
Checking rn

waruqi commented on xmake/rules/vala/xmake.lua:
```diff
@@ -58,31 +56,40 @@ rule("vala.build")
-        -- get c source file for vala
-        local sourcefile_c = target:autogenfile((sourcefile_vala:gsub(".vala$", ".c")))
-        local basedir = path.directory(sourcefile_c)
+        local argv = {"-C", "-d", target:autogendir()}
```
`-d`? why not `-b basedir`?

al1-ce replied:
- `-b` outputs all files into single directory
- `-d` outputs all files according to source tree
- 
We need `-d` because of these two lines

```lua
local sourcefile_c = target:autogenfile((sourcebatch:gsub(".vala$", ".c")))
local basedir = path.directory(sourcefile_c)
```

In this case for files in sub-directories `basedir` is going to be a sub-directory, for which `-b` flag doesn't account. It worked before because it was assumed that files were compiled one-by-one and directory for `-b` flag was a `basedir`

waruqi commented on xmake/rules/vala/xmake.lua:
```diff
@@ -58,31 +56,40 @@ rule("vala.build")
+        for _, sourcefile_vala in ipairs(sourcefiles_vala.sourcefiles) do
+            -- if it's only a vala file
+            if string.match(sourcefile_vala, ".vala$") ~= nil then
+                table.insert(argv, path(sourcefile_vala))
```
please add vala files list in argv tail

al1-ce commented:
Strange, vala/lua test doesn't compile even with upstream rule file...

al1-ce commented:
Implemented requested changes, all tests passing except for lua, but lua fails because of xmake (and maybe arch linux, it has headers who knows where) coz gcc fails to find lua includes from `-Llua` flag

Also fixed `vala.vapidir` and `vala.flags` not being applied and fixed `vala.vapifile` and `vala.header` locations being incorrect

waruqi commented on xmake/rules/vala/xmake.lua:
```diff
@@ -103,22 +99,48 @@ rule("vala.build")
+
+    on_buildcmd_file(function (target, batchcmds, sourcebatch, opt)
+        -- Again, only vala files need special treatment
+        if path.extension(sourcebatch) == ".vala" then
```
please rename `sourcebatch` to `sourcefile` in on_buildcmd_file

sourcebatch is only for on_buildcmd_files

waruqi commented on xmake/rules/vala/xmake.lua:
```diff
@@ -103,22 +99,48 @@ rule("vala.build")
+            end
+        end
+
+        batchcmds:show_progress(opt.progress, "${color.build.object}compiling.vala")
```
no sourcefile name in progress text. and Incremental compilation does not work.

```console
bash-5.2$ xmake
[ 25%]: compiling.vala
[100%]: build ok, spent 0.28s

bash-5.2$ xmake
[ 25%]: compiling.vala
[100%]: build ok, spent 0.28s
```

waruqi replied:
```console
$ cd staticlib

$ xmake
[ 20%]: compiling.vala
[ 40%]: archiving.release libmymath.a
[ 70%]: compiling.vala
[ 90%]: linking.release test
[100%]: build ok, spent 0.695s

$ xmake
[ 20%]: compiling.vala
[ 70%]: compiling.vala
[100%]: build ok, spent 0.517s
```

al1-ce replied:
Because, as I explained, valac has to be called each time, since it acts on all of `.vala` files

Same for lack of fileblname, it compiles all vala files together

waruqi replied:
you can show file progress one-by-one in for-loop.

al1-ce replied:
That would be something like this though:
```
[50%] compiling vala: src/main.vala
[50%] compiling vala: src/other.vala
[50%] compiling vala: src/sub/mod.vala
...
```
And so on for all sourcefiles displayed at once

If that's fine I'll correct it to do that

waruqi replied:
right, you can do it.

al1-ce replied:
Ok, I'll also try getting it to at least not recompile if no files been modified

al1-ce commented:
Couldn't get valac build to run only when needed, but everything else should be correct now

waruqi commented:
But incremental compilation still doesn't work.

```console
bash-5.2$ xmake
[ 20%]: compiling.vala src/mymath.vala
[ 70%]: compiling.vala src/main.vala
[100%]: build ok, spent 0.588s

bash-5.2$ xmake
[ 20%]: compiling.vala src/mymath.vala
[ 70%]: compiling.vala src/main.vala
[100%]: build ok, spent 0.571s
```

al1-ce commented:
Oh, I think I'm starting to understand how batchcmds work, not completely though...
Hard to figure out, not that much documentation about it, except for examples which do not explain anything

waruqi commented:
> Oh, I think I'm starting to understand how batchcmds work, not completely though... Hard to figure out, not that much documentation about it, except for examples which do not explain anything

https://xmake.io/#/manual/custom_rule?id=custom-batch-compile-script-process-one-source-file-at-a-time

al1-ce commented:
> https://xmake.io/#/manual/custom_rule?id=custom-batch-compile-script-process-one-source-file-at-a-time
Again, shows how, but now why. Maybe worth adding some explaining comments to it?

Anyway, thanks for taking time to review and merge my PR
